Libraries/xcexecution/Sources/NinjaExecutor.cpp

```diff
@@ -598,224 +670,185 @@ NinjaDependencyInfoExecutable()
-        for (std::string const &inputDependency : invocation.inputDependencies()) {
-            inputDependencies.push_back(ninja::Value::String(inputDependency));
-        }
+    /*
+     * Build up the bindings for the invocation.
+     */
+    std::vector<ninja::Binding> bindings = {
+        { "description", ninja::Value::String(description) },
+        { "dir", ninja::Value::String(Escape::Shell(invocation.workingDirectory())) },
+        { "exec", ninja::Value::String(exec) },
+    };
+    if (!environment.empty()) {
+        bindings.push_back({ "env", ninja::Value::String(environment) });
+    }
+    if (!dependencyInfoExec.empty()) {
+        bindings.push_back({ "depexec", ninja::Value::String(dependencyInfoExec) });
+    }
+    if (!dependencyInfoFile.empty()) {
+        bindings.push_back({ "depfile", ninja::Value::String(dependencyInfoFile) });
+    }
 
-        /*
-         * Build up order dependencies as literal Ninja values.
-         */
-        std::vector<ninja::Value> orderDependencies;
-        for (std::string const &orderDependency : invocation.orderDependencies()) {
-            orderDependencies.push_back(ninja::Value::String(orderDependency));
-        }
+    /*
+     * Build up outputs as literal Ninja values.
+     */
+    std::vector<ninja::Value> outputs;
+    for (std::string const &output : NinjaInvocationOutputs(invocation)) {
+        outputs.push_back(ninja::Value::String(output));
+    }
 
-        /*
-         * All invocations depend on the target containing them beginning.
-         */
-        orderDependencies.push_back(ninja::Value::String(targetBegin));
+    /*
+     * Build up inputs as literal Ninja values.
+     */
+    std::vector<ninja::Value> inputs;
+    for (std::string const &input : invocation.inputs()) {
+        inputs.push_back(ninja::Value::String(input));
+    }
 
-        /*
-         * Add the rule to build this invocation.
-         */
-        writer.build(outputs, NinjaRuleName(), inputs, bindings, inputDependencies, orderDependencies);
+    /*
+     * Build up input dependencies as literal Ninja values.
+     */
+    std::vector<ninja::Value> inputDependencies;
+    for (std::string const &inputDependency : invocation.inputDependencies()) {
+        inputDependencies.push_back(ninja::Value::String(inputDependency));
     }
 
     /*
-     * Serialize the Ninja file into the build root.
+     * Build up order dependencies as literal Ninja values.
      */
-    std::string path = TargetNinjaPath(target, targetEnvironment);
-    if (!WriteNinja(filesystem, writer, path)) {
-        fprintf(stderr, "error: unable to write target ninja: %s\n", path.c_str());
-        return false;
+    std::vector<ninja::Value> orderDependencies;
+    for (std::string const &orderDependency : invocation.orderDependencies()) {
+        orderDependencies.push_back(ninja::Value::String(orderDependency));
     }
 
+    /*
+     * All invocations depend on the target containing them beginning.
```

Comment:
There's also a third special rule added by this PR, that goes between a target's begin rule and the files in the target. That special rule is used as a barrier between writing out auxiliary files and beginning the target's file rules, in a similar way to the barriers between targets. I forget if that's added in the commit you're looking at here or the other one. :(